// dastoor/command-builder.js
function CommandBuilder() {
    this.commandStore = {};
}

CommandBuilder.prototype._get = function (name) {
    return this.commandStore[name];
};

CommandBuilder.prototype._set = function (command) {

    this.commandStore[command.path] = command;

    if(command.parentPath) {
        var parent = this.command(command.parentPath);
        parent.commands.push(command);
    }

    return command;
};

CommandBuilder.prototype.command = function (path) {
    return this._get(path) ||
           this._set(new Command(path));
};

var commandBuilder = new CommandBuilder();

function Command(path) {
    if(!/^[\w\-]+(\.[\w\-]+)*$/g.test(path)) throw 'module path is invalid: ' + path;

    var p = path.split('.');
    var name = p[p.length - 1];
    var parentPath = p.splice(0, p.length - 1).join('.');

    this.parentPath = parentPath;
    this.name = name;
    this.path = (parentPath ? parentPath + '.' : '') + name;
    this.help = {};
    this.commands = [];
    this.terminal = false;
    this.action = false;
}

Command.prototype.asTerminal = function () {
    this.terminal = true;
    return this;
};

Command.prototype._getChild = function (name) {
    return this.commands.filter(function (c) {
        return c.name === name;
    })[0];
};

Command.prototype.withHelp = function (options) {
    this.help = options;
    return this;
};

Command.prototype.withAction = function (action) {
    this.action = action;
    return this;
};

module.exports = commandBuilder;
