// dastoor/runner.js
var minimist = require('minimist');
var extend = require('extend');
var utils = require('./_utils');

var defaultOptions = {
    log: console.log,
    errorLog: console.log,
    helpLog: console.log,
    shortHelpKey: 'h',
    longHelpKey: 'help',
    helpKeyDescription: 'show this help'
};

function Runner(options) {
    this.options = extend({}, defaultOptions, options);
    this.onModuleLoadRegistrar = [];
}

Runner.prototype._isHelpKey = function(e) {
    return e === ('--' + this.options.longHelpKey) ||
           e === ('-' + this.options.shortHelpKey);
};

Runner.prototype.run = function (entry, args) {
    var e,
        path = [];

    while(args.length){
        e = args[0];
        if(this._isHelpKey(e)) return this.renderHelp(root, entry);
        entry = entry._getChild(e);
        if(!entry) break;
        if(entry.terminal) break;

        path.push(e);
        args = args.splice(1);
    }

    if(entry){
        var hasHelp = args.filter(this._isHelpKey.bind(this)).length > 0;
        var controller = entry.controller;

        if(hasHelp || !controller) return this.renderHelp(root, entry);

        path.push(e);
        this.onModuleLoadRegistrar.forEach(function (f) {
            f.call(entry, entry, path);
        });

        var inputArgs = minimist(args.splice(1));
        extend(inputArgs, this.options.localArgs || {});
        controller.call(entry, inputArgs);
    }
    else{
        this.options.errorLog('command `'+e+'` not found');
    }
};

Runner.prototype.onModuleLoad = function (f) {
    this.onModuleLoadRegistrar.push(f);
};

Runner.prototype.renderHelp = function (root, command) {
    var helpOption = {
        name: '-'+this.options.shortHelpKey+', --' + this.options.longHelpKey,
        description: this.options.helpKeyDescription
    };

    var log = this.options.helpLog;

    var h = command.helpDetails;
    var sub = command.commands;
    var usage = h && h.usage;
    var options = ((h && h.options) || []).concat(helpOption);
    log();
    if(h && h.description){
        log('    ' + h.description);
        log();
    }

    if(options){
        log('    Options:');
        options.forEach(function (o) {
            log('      ' + utils.pad(o.name, 20), o.description);
        });
        log();
    }

    if(usage || sub.length > 0){
        log('    Usage:');
        if(usage){
            usage.forEach(function (u) {
                log('      ' + u);
            });
        }

        if(sub.length > 0){
            sub.forEach(function (subcmd) {
                var p = subcmd.name.replace(/\./g, ' ');
                var h = subcmd.helpDetails || {};
                log('      '+root.name+' ' + utils.pad(p, 20), h.description || '');
            });
            log();
        }
    }
};

module.exports = Runner;
